flink-streaming-connectors/flink-connector-hbase/src/main/java/org/apache/flink/streaming/connectors/hbase/MutationActionList.java

```diff
@@ -0,0 +1,371 @@
+	public List<Mutation> newMutationList(byte[] rowKey, boolean writeToWAL) {
+		List<Mutation> mutations = new ArrayList<>();
+		Put put = null;
+		Increment increment = null;
+		Append append = null;
+		Delete delete = null;
+		boolean rowIsDeleted = false;
+		for (MutationAction action : actions) {
+			switch (action.getType()) {
+				case PUT:
+					if (put == null) {
+						put = new Put(rowKey);
+						mutations.add(put);
+					}
+					if (action.getTs() == -1) {
+						put.addColumn(action.getFamily(), action.getQualifier(), action.getValue());
+					} else {
+						put.addColumn(action.getFamily(), action.getQualifier(), action.getTs(), action.getValue());
+					}
+					break;
+
+				case INCREMENT:
+					if (increment == null) {
+						increment = new Increment(rowKey);
+						mutations.add(increment);
+					}
+					increment.addColumn(action.getFamily(), action.getQualifier(), action.getIncrement());
+					break;
+
+				case APPEND:
+					if (append == null) {
+						append = new Append(rowKey);
+						mutations.add(append);
+					}
+					append.add(action.getFamily(), action.getQualifier(), action.getValue());
+					break;
+
+				// If there are multiple DELETE_ROW actions, only the first one is served
+				case DELETE_ROW:
+					if (!rowIsDeleted) {
+						for (int i = 0; i < mutations.size(); ) {
+							if (mutations.get(i) instanceof Delete) {
+								mutations.remove(i);
+							} else {
+								i++;
+							}
+						}
+						delete = new Delete(rowKey, action.getTs());
+						mutations.add(delete);
+						rowIsDeleted = true;
+					}
+					break;
+
+				case DELETE_FAMILY:
+					if (!rowIsDeleted) {
+						if (delete == null) {
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addFamily(action.getFamily(), action.getTs());
+					}
+					break;
+
+				case DELETE_COLUMNS:
+					if (!rowIsDeleted) {
+						if (delete == null) {
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addColumns(action.getFamily(), action.getQualifier(), action.getTs());
+					}
+					break;
+
+				case DELETE_COLUMN:
+					if (!rowIsDeleted) {
+						if (delete == null) {
+							delete = new Delete(rowKey);
+							mutations.add(delete);
+						}
+						delete.addColumn(action.getFamily(), action.getQualifier(), action.getTs());
+					}
+					break;
+
+				default:
+					throw new IllegalArgumentException("Cannot process such action type: " + action.getType());
+			}
+		}
+		Durability durability = writeToWAL ? Durability.SYNC_WAL : Durability.SKIP_WAL;
+		for (Mutation mutation : mutations) {
+			mutation.setDurability(durability);
+		}
+		return mutations;
+	}
+
+	/**
+	 * Create a new list of HBase {@link Mutation}s enabling WAL as default.
+	 *
+	 * @param rowKey row that the created {@link Mutation} list is applied to
+	 * @return a list of HBase {@link Mutation}s
+	 */
+	public List<Mutation> newMutationList(byte[] rowKey) {
+		return newMutationList(rowKey, true);
```

Comment:
> And would it be even better to make newMutationList(byte[] rowKey, boolean writeToWAL) a non-public method?

Exactly. That was my point. My below comments also was to make things cleaner. So MutationActionList is exposed to the user. It should only have getMutations() or createMutations() and addXXX(). So when the user creates a MuationActionList using addXXX APIs in it. Then he should do getMutations() or createMutations()(add suitable javadoc) saying this API gives a list of mutations that can be sent to HBase. In this method you need to do the logic of what newMutationList() does. Do not expose getActions() also. I was suggesting to do this in MutationAction but fine in MutationActionList as only that is going to be exposed now. Better rename it to MutationActions. Mark MutationAction inner class Private. And none of those Getters in the inner class should be exposed to User. If needed you can create it as a separate class also.
And one important thing, the rowKey() in HBaseMapper decides the row for which these mutations are getting created. If this is set to null then there is no MutationActionList possible at all. So you should ensure that the rowKey if null then the mapper should fail. If not just don''t add any rowKey explicitly - Anyway actions(IN value) API  accepts the rowkey that should be enough. Inside the impl - if the value is found null throw exception and add a javadoc about it. It should be enough and will be simple too.